src/translate.c

```diff
@@ -110,87 +113,90 @@ static r_ssize chr_find_normalize_start(SEXP x, r_ssize size) {
-      r_ssize elt_size = r_length(elt);
-      r_ssize elt_start = chr_find_normalize_start(elt, elt_size);
+  r_ssize size = r_length(x);
+  const SEXP* p_x = VECTOR_PTR_RO(x);
 
-      if (elt_size == elt_start) {
-        break;
-      }
+  for (r_ssize i = 0; i < size; ++i) {
+    SEXP elt_old = p_x[i];
 
-      elt = chr_normalize_encoding(elt, elt_size, elt_start);
-      SET_VECTOR_ELT(x, i, elt);
-      break;
+    SEXP elt_new = obj_normalize_encoding(elt_old);
+    if (elt_old == elt_new) {
+      continue;
     }
-    case VECSXP: {
-      r_ssize elt_size = r_length(elt);
-      r_ssize elt_start = list_find_normalize_start(elt, elt_size);
-
-      if (elt_size == elt_start) {
-        break;
-      }
+    PROTECT(elt_new);
 
-      elt = list_normalize_encoding(elt, elt_size, elt_start);
-      SET_VECTOR_ELT(x, i, elt);
-      break;
-    }
-    default:
-      continue;
+    if (MAYBE_REFERENCED(x)) {
+      x = PROTECT(r_clone_referenced(x));
+      ++nprot;
+      p_x = VECTOR_PTR_RO(x);
     }
+
+    SET_VECTOR_ELT(x, i, elt_new);
+    UNPROTECT(1);
   }
 
-  UNPROTECT(1);
+  UNPROTECT(nprot);
   return x;
 }
 
-static inline bool elt_all_normalized(SEXP x);
+// -----------------------------------------------------------------------------
 
-static r_ssize list_find_normalize_start(SEXP x, r_ssize size) {
-  const SEXP* p_x = VECTOR_PTR_RO(x);
+static SEXP attrib_normalize_encoding(SEXP x);
 
-  for (r_ssize i = 0; i < size; ++i) {
-    const SEXP elt = p_x[i];
+static
+SEXP obj_attrib_normalize_encoding(SEXP x) {
+  SEXP attrib_old = r_attrib(x);
 
-    if (elt_all_normalized(elt)) {
-      continue;
-    }
+  if (attrib_old == r_null) {
+    return x;
+  }
 
-    return i;
+  SEXP attrib_new = attrib_normalize_encoding(attrib_old);
+  if (attrib_new == attrib_old) {
+    return x;
   }
+  PROTECT(attrib_new);
 
-  return size;
-}
+  x = PROTECT(r_clone_referenced(x));
+  r_poke_attrib(x, attrib_new);
 
-static inline bool elt_all_normalized(SEXP x) {
-  switch (TYPEOF(x)) {
-  case STRSXP: {
-    r_ssize size = r_length(x);
-    r_ssize start = chr_find_normalize_start(x, size);
-    return size == start;
-  }
-  case VECSXP: {
-    r_ssize size = r_length(x);
-    r_ssize start = list_find_normalize_start(x, size);
-    return size == start;
-  }
-  default: {
-    return true;
-  }
-  }
+  UNPROTECT(2);
+  return x;
 }
 
-// -----------------------------------------------------------------------------
+static
+SEXP attrib_normalize_encoding(SEXP x) {
+  int nprot = 0;
+  r_ssize loc = 0;
 
-// For testing
-// [[ register() ]]
-SEXP vctrs_normalize_encoding(SEXP x) {
-  return vec_normalize_encoding(x);
+  for (SEXP node = x; node != r_null; node = r_node_cdr(node), ++loc) {
+    SEXP elt_old = r_node_car(node);
+
+    SEXP elt_new = obj_normalize_encoding(elt_old);
+    if (elt_old == elt_new) {
+      continue;
+    }
+    PROTECT(elt_new);
+
+    if (MAYBE_REFERENCED(x)) {
+      x = PROTECT(r_clone_referenced(x));
```

Comment:
The referenced check is not correct here since tails of pairlists can be reused by fresh nodes.

We only need to duplicate the pairlist nodes up to `loc` and we could reuse the tail of the pairlist. Then recurse back into `attrib_normalise_encoding()` from the `loc`th node (or equivalently update the state and `continue;`, that's probably better). That would avoid cloning and traversing the attributes multiple times when multiple attributes need normalisation.